File: Homework/Quang_Minh/Lesson3/KClosetPointsToOrigin.java
package HomeWorkLesson3;

public class KClosetPointsToOrigin {
    public int[][] kClosest(int[][] points, int k) {
        int[][] result = new int[k][2];
        quickSelect(points, k-1, 0, points.length-1);
        for (int i = 0 ; i < k ; i++){
            result[i] = points[i];
        }
        return result;
    }
    public static int[] quickSelect(int[][] nums, int k, int start, int end) {
        if (nums.length == 1) return nums[0];
        int indexPivot = partition(nums, start, end);
        if (indexPivot == k) return nums[indexPivot];
        if (indexPivot > k) return quickSelect(nums, k, start, indexPivot - 1);
        return quickSelect(nums, k, indexPivot + 1, end);
    }

    public static int partition(int[][] array, int start, int end) {
        int[] random = array[end];
        int position = start;
        for (int i = start; i < end; i++) {
            if (getDistance(array[i]) <= getDistance(random)) {
                swap(array,i,position);
                position++;
            }
        }
        swap(array,position,end);
        return position;
    }
    public static void swap(int[][] array, int i, int j) {
        int[] temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }
    public static double getDistance(int[]input){
        return Math.pow(input[0],2) + Math.pow(input[1] , 2);
    }
}

File: Homework/Van_Le/lesson3/KClosetPointsToOrigin/KClosetPointsToOrigin.cs
using System;

namespace KClosetPointsToOrigin
{
    public class KClosetPointsToOrigin
    {
        public int[][] KClosest(int[][] points, int k)
        {
            return QuickSelect(points, k);
        }
        private int[][] QuickSelect(int[][] points, int k)
        {
            int[][] result = new int[k][];
            int left = 0, right = points.Length - 1;
            int pivotIndex = points.Length;
            while (pivotIndex != k)
            {
                pivotIndex = Partition(points, left, right);
                if (pivotIndex < k)
                {
                    left = pivotIndex;
                }
                else
                {
                    right = pivotIndex - 1;
                }
            }

            Array.Copy(points, 0, result, 0, k);
            return result;
        }

        private int Partition(int[][] points, int left, int right)
        {
            int[] pivot = ChoosePivot(points, left, right);
            int pivotDist = SquaredDistance(pivot);
            while (left < right)
            {
                if (SquaredDistance(points[left]) >= pivotDist)
                {
                    int[] temp = points[left];
                    points[left] = points[right];
                    points[right] = temp;
                    right--;
                }
                else
                {
                    left++;
                }
            }

            if (SquaredDistance(points[left]) < pivotDist)
                left++;
            return left;
        }

        private int[] ChoosePivot(int[][] points, int left, int right)
        {
            return points[left + (right - left) / 2];
        }

        private int SquaredDistance(int[] point)
        {
            return point[0] * point[0] + point[1] * point[1];
        }
    }
}
